src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,426 @@
+                    Debug.Assert(_numPendingRemoves <= _pendingRemoves.Length);
+                    Debug.Assert(_numUserWaits >= 0);
+                    Debug.Assert(_numUserWaits <= _registeredWaits.Length);
+                    Debug.Assert(_numPendingRemoves <= _numUserWaits, $"Num removals {_numPendingRemoves} should be less than or equal to num user waits {_numUserWaits}");
+                    
+                    if (_numPendingRemoves == 0 || _numUserWaits == 0)
+                    {
+                        return;
+                    }
+                    int originalNumUserWaits = _numUserWaits;
+                    int originalNumPendingRemoves = _numPendingRemoves;
+
+                    // This is O(N^2), but max(N) = 63 and N will usually be very low
+                    for (int i = 0; i < _numPendingRemoves; i++)
+                    {
+                        for (int j = 0; j < _numUserWaits; j++)
+                        {
+                            if (_pendingRemoves[i] == _registeredWaits[j])
+                            {
+                                _registeredWaits[j].TrySignalUserWaitHandle();
+                                _registeredWaits[j] = _registeredWaits[_numUserWaits - 1];
+                                _waitHandles[j + 1] = _waitHandles[_numUserWaits];
+                                _registeredWaits[_numUserWaits - 1] = null;
+                                _waitHandles[_numUserWaits] = null;
+                                --_numUserWaits;
+                                _pendingRemoves[i] = null;
+                                break;
+                            }
+                        }
+                        Debug.Assert(_pendingRemoves[i] == null);
+                    }
+                    _numPendingRemoves = 0;
+
+                    Debug.Assert(originalNumUserWaits - originalNumPendingRemoves == _numUserWaits,
+                        $"{originalNumUserWaits} - {originalNumPendingRemoves} == {_numUserWaits}");
+                }
+                finally
+                {
+                    _registeredHandlesLock.Release();
+                    _removesLock.Release();
+                }
+            }
+            
+            /// <summary>
+            /// Queue a call to <see cref="CompleteWait(object)"/> on the ThreadPool.
+            /// </summary>
+            /// <param name="registeredHandle">The handle that completed.</param>
+            /// <param name="timedOut">Whether or not the wait timed out.</param>
+            private void QueueWaitCompletion(RegisteredWaitHandle registeredHandle, bool timedOut)
+            {
+                registeredHandle.RequestCallback();
+                if (registeredHandle.Repeating)
+                {
+                    registeredHandle.RestartTimeout(Environment.TickCount);
+                }
+                else
+                {
+                    UnregisterWait(registeredHandle, blocking: false);
```

Comment:
Talked offline. Current design works. The comment in `UnregisterWait(RegisteredWaitHandle,bool)` explains why.